persist/fs/mmap_linux.go

```diff
@@ -56,22 +56,48 @@ func mmap(fd, offset, length int64, flags int, opts mmapOptions) ([]byte, error)
 		prot = prot | syscall.PROT_WRITE
 	}
 
-	b, err := syscall.Mmap(int(fd), offset, int(length), prot, flags)
-	if err != nil {
-		return nil, fmt.Errorf("mmap error: %v", err)
+	flagsWithoutHugeTLB := flags
+	shouldUseHugeTLB := opts.hugeTLB.enabled && length >= opts.hugeTLB.threshold
+	if shouldUseHugeTLB {
+		// We use the MAP_HUGETLB flag instead of MADV_HUGEPAGE because transparent
+		// hugepages only work with anonymous, private pages. Please see the MADV_HUGEPAGE
+		// section of http://man7.org/linux/man-pages/man2/madvise.2.html and the MAP_HUGETLB
+		// section of http://man7.org/linux/man-pages/man2/mmap.2.html for more details.
+		flags = flags | syscall.MAP_HUGETLB
 	}
 
-	if !opts.hugePages.enabled ||
-		length < opts.hugePages.threshold {
-		return b, nil
+	var (
+		b          []byte
+		err        error
+		withTLBErr error
+		warning    error
+	)
+	b, err = syscall.Mmap(int(fd), offset, int(length), prot, flags)
+	// Save in case we need to include it in the warning later
+	withTLBErr = err
+	// Sometimes allocations that specify huge pages will fail because the O.S
+	// isn't configured properly or there are not enough available huge pages in
+	// the pool. You can try and allocate more by executing:
+	// 		echo 20 > /proc/sys/vm/nr_hugepages
+	// See this document for more details: https://www.kernel.org/doc/Documentation/vm/hugetlbpage.txt
+	// Regardless, we don't want to fail hard in that scenario. Instead, we try
+	// and mmap without the hugeTLB flag.
+	if err != nil && shouldUseHugeTLB {
+		b, err = syscall.Mmap(int(fd), offset, int(length), prot, flagsWithoutHugeTLB)
+		// If we succeeded the second time, then proceed but make sure the caller
+		// receives a warning that includes the error from when we tried to use the
+		// hugeTLB flag.
+		if err == nil {
+			warning = fmt.Errorf(
+				"error while trying to mmap with hugeTLB flag: %s, hugeTLB disabled", withTLBErr.Error())
+		}
 	}
 
-	// Reduce the number of pagefaults when scanning through large files
-	if err := syscall.Madvise(b, syscall.MADV_HUGEPAGE); err != nil {
-		return nil, fmt.Errorf("madvise error, check platform support: %v", err)
+	if err != nil {
+		return mmapResult{}, fmt.Errorf("mmap error: %v", err)
 	}
 
-	return b, nil
+	return mmapResult{result: b}, nil
```

Comment:
Missing actually setting the warning here, should this be `mmapResult{result: b, warning: warning}`?

Reply:
yep, fixed